### PR TITLE
zh-cn: init translation of svelte variables props

### PR DESCRIPTION
**Pull request by hanyujie2002**
### Description

Same as title
### Motivation

Ditto
### Additional details

None
### Related issues and pull requests

Relates to #16965

### CODE REVIEW

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md [hunk 1]
```diff
@@ -0,0 +1,487 @@
...
+    newTodoId = Math.max(...todos.map((t) => t.id)) + 1;
+  }
+}
+```
+
+> **备注：** 如你所见，响应式陈述不仅限于单行。以下代码也可以生效，但可读性较差：`$: newTodoId = totalTodos ? Math.max(...todos.map((t) => t.id)) + 1 : 1`
+
+2. Svelte 是如何做到这一点的？编译器会解析整个响应式陈述并检测到它依赖于 `totalTodos` 变量和 `todos` 数组。所以每当它们中任何一个被修改时，都会重新评估代码并相应地更新 `newTodoId`。
+
+让我们在 `addTodo()` 函数中使用它。更新它如下：
```
Comment (hanyujie2002) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md:368]:
```suggestion
   让我们在 `addTodo()` 函数中使用它。更新它如下：
```

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md [hunk 2]
```diff
@@ -0,0 +1,487 @@
...
+
+```js
+function addTodo() {
+  todos = [...todos, { id: newTodoId, name: newTodoName, completed: false }];
+  newTodoName = "";
+}
+```
```
Comment (hanyujie2002) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md:375]:
```suggestion
   ```js
   function addTodo() {
     todos = [...todos, { id: newTodoId, name: newTodoName, completed: false }];
     newTodoName = "";
   }
```
```

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+   现在让我们用这些信息做点什么。
+
+2. 让我们从显示状态信息开始。找到 `id` 为 `list-heading` 的 `<h2>` 标题并用动态表达式替换已经硬编码的当前任务及已完成任务数量：
+
+   ```svelte
+   <h2 id="list-heading">{completedTodos} out of {totalTodos} items completed</h2>
+   ```
+
+3. 回到应用程序，你应该会看到和以前一样的“2 out of 3 items completed”信息，但这次的信息来自 `todos` 数组。
+4. 为了证明这一点，回到数组并尝试改变一些待办事项对象的已完成属性值，甚至是新增一个新的待办事项对象。观察信息中的数字是如何恰当地被更新。
```
Comment (jasonren0403) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md:99]:
```suggestion
4. 为了证明这一点，回到数组并尝试改变一些待办事项对象的已完成属性值，甚至是新增一个新的待办事项对象。观察信息中的数字是如何被恰当地更新的。
```

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+
+## 从数据动态地生成待办事项
+
+目前我们显示的待办事项都是静态的。而我们想要遍历 `todos` 数组中的每个项目并为每个任务渲染其标记，让我们马上动手吧。
+
+HTML 没有表达逻辑的方式（像是条件和循环）。但 Svelte 做到了。在这种情况下，我们使用 [`{#each}`](https://svelte.dev/docs#each) 指令来迭代 `todos` 数组。假如有提供第二个参数，将包含当前项目的索引。此外，可以提供一个关键表达式，它将唯一识别每个项目。Svelte 将在数据改变时使用它来区分清单项目，而不是在最后新增或删除项目，所以始终指定关键表达式是一种好的做法。最后，可以提供 `:else` 区块，它将在清单为空时呈现。
```
Comment (jasonren0403) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md:105]:
```suggestion
HTML 没有表达逻辑的方式（如条件和循环），但 Svelte 可以做到。在这种情况下，我们使用 [`{#each}`](https://svelte.dev/docs#each) 指令来迭代 `todos` 数组。假如提供了第二个参数，将包含当前项目的索引。此外，可以提供一个键表达式，它将唯一识别每个项目。当数据发生变化时，Svelte 将使用它来更改列表，而不是在末尾添加或删除项目，所以始终指定键表达式是一种好的做法。最后，可以提供 `:else` 区块，它将在列表为空时渲染如此。
```

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+
+1. 在 `Todos.svelte` 中，将现有的 `let todos = …` 区块替换为 `export let todos = []`。
+
+   ```js
+   export let todos = [];
+   ```
+
+   起初这可能会让你觉得有点奇怪。这不像在 JavaScript 模块中使用 `export` 的常见方式！但这就是 Svelte 通过采用赋予有效语法其新用途来“扩展”JavaScript 的做法。在这种情况下，Svelte 使用 `export` 关键字将变量声明标记为属性，这意味着组件的消费者可以访问它。
+
+   你也可以为属性指定默认初始值。如果组件的消费者没有在组件上指定属性时，将使用此默认初始值——或者如果它的初始值未定义——在实例化组件时。
```
Comment (jasonren0403) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md:174]:
```suggestion
   你也可以为属性指定默认初始值。如果组件的消费者没有在组件上指定属性，或者在实例化组件时属性的初始值未定义时，则会使用该初始值。
```

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md [hunk 6]
```diff
@@ -0,0 +1,488 @@
...
+       { id: 1, name: "Create a Svelte starter app", completed: true },
+       { id: 2, name: "Create your first component", completed: true },
+       { id: 3, name: "Complete the rest of the tutorial", completed: false }
+     ];
+   </script>
+
+   <Todos todos={todos} />
+   ```
+
+4. 当属性和变量具有相同名称时，Svelte 允许你仅指定变量作为便捷的捷径，因此我们可以像这样重写最后一行。现在试试吧。
```
Comment (jasonren0403) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md:195]:
```suggestion
4. 当属性和变量具有相同名称时，Svelte 允许你仅指定变量作为捷径，因此我们可以像这样重写最后一行。现在试试吧。
```

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md [hunk 7]
```diff
@@ -0,0 +1,488 @@
...
+   </button>
+   ```
+
+   在 Svelte 中处理器的一个常见错误是将执行函数的结果作为处理器传递，而不是传递函数。例如，如果你指定 `on:click={removeTodo(todo)}`，它将执行 `removeTodo(todo)` 并将结果作为处理器传递，但这不是我们想要的。
+
+   在这种情况下，你必须指定 `on:click={() => removeTodo(todo)}` 作为处理器。如果 `removeTodo()` 没有接收参数，你可以使用 `on:event={removeTodo}`，但不能使用 `on:event={removeTodo()}`。这不是 Svelte 中的特殊语法——在这里我们只是使用常规的 JavaScript [箭头函数](/zh-CN/docs/Web/JavaScript/Reference/Functions/Arrow_functions)。
+
+这是不错的进展——现在我们可以删除任务了。当待办事项的*Delete*按钮被按下时，相关的待办事项将从 `todos` 数组中移除，并更新用户界面以不再显示它。此外，我们现在可以勾选复选框，相关待办事项的已完成状态将在 `todos` 数组中被更新。
+
+然而，“x out of y items completed”标头并未被更新。继续阅读以了解为什么会出现这种情况以及我们如何解决它。
```
Comment (jasonren0403) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md:240]:
```suggestion
然而，“x out of y items completed”标题并未被更新。请继续阅读，了解出现这种情况的原因和解决方法。
```

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md [hunk 8]
```diff
@@ -0,0 +1,488 @@
...
+```js
+let totalTodos = todos.length;
+let completedTodos = todos.filter((todo) => todo.completed).length;
+```
+
+我们可以在勾选和移除待办事项后重新计算它们，但有一种更简单的方式可以做到。
+
+可以通过在前面加上 `$:` 前缀，告诉 Svelte 我们希望 `totalTodos` 和 `completedTodos` 等变量具有响应性。Svelte 将生成代码以在它们依赖的数据发生变化时自动更新它们。
+
+> **备注：** Svelte 使用 `$:` [JavaScript 标记陈述语法](/zh-CN/docs/Web/JavaScript/Reference/Statements/label)来标记响应性陈述。就像用于声明属性的 `export` 关键字一样，这可能看起来有点陌生。但这是 Svelte 利用有效的 JavaScript 语法并赋予其新用途的另一个例子——在这种情况下，意味着“每当任何引用的值发生变化时，就会重新执行此代码”。一旦习惯了，你就无法回头了。
```
Comment (jasonren0403) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md:257]:
```suggestion
> **备注：** Svelte 使用 `$:` [JavaScript 标记语句语法](/zh-CN/docs/Web/JavaScript/Reference/Statements/label)来标记响应性语句。就像用于声明属性的 `export` 关键字一样，这可能看起来有点陌生。但这是 Svelte 利用有效的 JavaScript 语法并赋予其新用途的另一个例子——在这种情况下，意味着“每当任何引用的值发生变化时，就会重新执行此代码”。一旦习惯了，你就无法回头了。
```

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md [hunk 9]
```diff
@@ -0,0 +1,488 @@
...
+更新 `src/components/Todos.svelte` 中的 `totalTodos` 和 `completedTodos` 变量定义，如下：
+
+```js
+$: totalTodos = todos.length;
+$: completedTodos = todos.filter((todo) => todo.completed).length;
+```
+
+如果你现在检查你的应用程序，当待办事项完成或被删除时，你将会看到标题的数字被更新。做得好！
+
+Svelte 编译器在幕后会解析和分析我们的代码以生成依赖树，然后它会产生 JavaScript 代码以在其中一个依赖项被更新时，重新评估每个响应性陈述。Svelte 的响应性以非常轻量和高性能的方式来实现，无需使用监听器（listener）、设置器（setter）、获取器（getter）或任何其他复杂的机制。
```
Comment (jasonren0403) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md:268]:
```suggestion
在幕后，Svelte 编译器会解析和分析我们的代码，生成依赖关系树，然后生成 JavaScript 代码，以便在依赖关系更新时重新评估每个响应式语句。Svelte 中的响应性是以非常轻量级和高性能的方式实现的，无需使用监听器（listener）、设置器（setter）、获取器（getter）或任何其他复杂机制。
```

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md [hunk 10]
```diff
@@ -0,0 +1,488 @@
...
+
+当前文章的下一个主要任务是添加一些用来添加新的待办事项的功能。
+
+1. 首先，我们将创建一个变量来保存新待办事项的文字。将此声明添加到 `Todos.svelte` 文件的 `<script>` 区块中：
+
+   ```js
+   let newTodoName = "";
+   ```
+
+2. 我们将使用 `<input>` 中的值来添加新任务。为此，我们需要将 `newTodoName` 变量绑定到 `todo-0` 输入框，以使 `newTodoName` 变量的值与输入框的 `value` 属性保持同步。我们可以这样做：
```
Comment (jasonren0403) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md:280]:
```suggestion
2. 现在，我们将在 `<input>` 中使用此值来添加新任务。为此，我们需要将 `newTodoName` 变量绑定到 `todo-0` 输入框，以使 `newTodoName` 变量的值与输入框的 `value` 属性保持同步。我们可以这样做：
```

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md [hunk 11]
```diff
@@ -0,0 +1,488 @@
...
+   ```js
+   function addTodo() {
+     todos = [...todos, { id: 999, name: newTodoName, completed: false }];
+     newTodoName = "";
+   }
+   ```
+
+## 给每个待办事项唯一的 ID
+
+如果现在尝试在你的应用程序中添加新的待办事项，你将可以添加新的待办事项并让它出现在用户界面中——一次。但如果你尝试第二次，它将没有效果，并且你将收到一个控制台消息“Error: Cannot have duplicate keys in a keyed each”（错误：不能有重复的键）。看起来我们的待办事项需要唯一 ID。
```
Comment (jasonren0403) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md:350]:
```suggestion
如果现在尝试在你的应用程序中添加新的待办事项，你将可以添加新的待办事项并让它出现在用户界面中——一次。但如果你尝试第二次，它将没有效果，并且你将收到一个控制台消息“Error: Cannot have duplicate keys in a keyed each”（错误：不能有重复的键）。我们的待办事项需要唯一 ID。
```

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md [hunk 12]
```diff
@@ -0,0 +1,488 @@
...
+       newTodoId = 1;
+     } else {
+       newTodoId = Math.max(...todos.map((t) => t.id)) + 1;
+     }
+   }
+   ```
+
+   > **备注：** 如你所见，响应式陈述不仅限于单行。以下代码也可以生效，但可读性较差：`$: newTodoId = totalTodos ? Math.max(...todos.map((t) => t.id)) + 1 : 1`
+
+2. Svelte 是如何做到这一点的？编译器会解析整个响应式陈述并检测到它依赖于 `totalTodos` 变量和 `todos` 数组。所以每当它们中任何一个被修改时，都会重新评估代码并相应地更新 `newTodoId`。
```
Comment (jasonren0403) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md:367]:
```suggestion
2. Svelte 是如何做到这一点的？编译器会解析整个响应式语句并检测到它依赖于 `totalTodos` 变量和 `todos` 数组。所以每当它们中任何一个被修改时，都会重新评估代码并相应地更新 `newTodoId`。
```

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md [hunk 13]
```diff
@@ -0,0 +1,488 @@
...
+     filter === "active"
+       ? todos.filter((t) => !t.completed)
+       : filter === "completed"
+         ? todos.filter((t) => t.completed)
+         : todos;
+   ```
+
+   我们使用 `filter` 变量来控制当前的筛选器：_all_、_active_、_completed_。只需将其中一个值赋给筛选器变量，即可使用该筛选器并更新待办事项列表。让我们看看它是如何实现的。
+
+   ``filterTodos()` 函数接收当前的筛选器和待办事项列表，并返回经过筛选的新待办事项数组。
```
Comment (jasonren0403) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md:396]:
```suggestion
   `filterTodos()` 函数接收当前的筛选器和待办事项列表，并返回经过筛选的新待办事项数组。
```

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md [hunk 14]
```diff
@@ -0,0 +1,488 @@
...
+       <span class="visually-hidden">Show</span>
+       <span>Completed</span>
+       <span class="visually-hidden">tasks</span>
+     </button>
+   </div>
+   ```
+
+   这段标记中发生了几件事情。
+
+   我们通过将 `btn__primary` 类应用于筛选器按钮来显示当前使用的筛选器。为了有条件地将样式类应用于元素上，我们使用 `class:name={value}` 指令。如果值表达式的计算结果为真，那么类名将被应用。你可以将许多不同条件的指令放在同一个元素上。所以当我们写为 `class:btn__primary={filter === 'all'}` 时，如果筛选器（filter）变量等于 all，Svelte 将会应用 `btn__primary` 类。
```
Comment (jasonren0403) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md:422]: apply code format

```suggestion
   我们通过将 `btn__primary` 类应用于筛选器按钮来显示当前使用的筛选器。为了有条件地将样式类应用于元素上，我们使用 `class:name={value}` 指令。如果值表达式的计算结果为真，那么类名将被应用。你可以将许多不同条件的指令放在同一个元素上。所以当我们写为 `class:btn__primary={filter === 'all'}` 时，如果筛选器（filter）变量的值等于 `'all'`，Svelte 将会应用 `btn__primary` 类。
```

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md [hunk 15]
```diff
@@ -0,0 +1,488 @@
...
+
+   > **备注：** Svelte 提供了一个便利的快捷方式，允许我们在类与变量名称相同时，可以将 `<div class:active={active}>` 缩短为 `<div class:active>`。
+
+   ``aria-pressed={filter === 'all'}` 也是类似的情况：当在大括号之间传入的 JavaScript 表达式的计算结果为真时，`aria-pressed` 属性将被加入到按钮上。
```
Comment (jasonren0403) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md:426]:
```suggestion
   `aria-pressed={filter === 'all'}` 也是类似的情况：当在大括号之间传入的 JavaScript 表达式的计算结果为真值时，`aria-pressed` 属性将被加入到按钮上。
```

**Review by jasonren0403 — CHANGES_REQUESTED** [review]
some suggestions

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md (hunk 5 above)
Comment (hanyujie2002) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_variables_props/index.md:174]:
```suggestion
   你也可以为属性指定默认初始值。如果组件的消费者没有在组件上指定属性，或者在实例化组件时属性的初始值未定义，则会使用该初始值。
```

**Review by jasonren0403 — APPROVED** [review]
LGTM, Thank you @hanyujie2002 👍

**Pull request merged**